resources/aws.go

```diff
@@ -0,0 +1,334 @@
+func (obj *AwsRes) Init() error {
+	return obj.BaseRes.Init() // call base init, b/c we're overriding
+}
+
+// Watch is the primary listener for this resource and it outputs events.
+func (obj *AwsRes) Watch() error {
+	send := false
+	var exit *error
+	sess := session.Must(session.NewSession(&aws.Config{
+		Region: aws.String(obj.Region),
+	}))
+	svc := ec2.New(sess)
+	diInput := ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("tag:Name"),
+				Values: []*string{aws.String(obj.GetName())},
+			},
+			{
+				Name:   aws.String("instance-state-name"),
+				Values: []*string{aws.String("running"), aws.String("pending"), aws.String("stopped"), aws.String("stopping")},
+			},
+		},
+	}
+	if err := obj.Running(); err != nil {
+		return err
+	}
+	awschan := make(chan string)
+	go func() {
+		for {
+			diOutput, err := svc.DescribeInstances(&diInput)
+			if err != nil {
+				errwrap.Wrapf(err, "error describing instances")
+			}
+			var waitInput ec2.DescribeInstancesInput
+			if len(diOutput.Reservations) == 1 {
+				waitInput = ec2.DescribeInstancesInput{
+					InstanceIds: []*string{diOutput.Reservations[0].Instances[0].InstanceId},
+				}
+			} else {
+				waitInput = diInput
+			}
+
+			if obj.State == "running" && len(diOutput.Reservations) == 1 {
+				svc.WaitUntilInstanceStopped(&waitInput)
+				if *diOutput.Reservations[0].Instances[0].State.Name != "running" {
+					awschan <- "stopped"
+					svc.WaitUntilInstanceRunning(&waitInput)
+				}
+			}
+			if obj.State == "stopped" && len(diOutput.Reservations) == 1 {
+				svc.WaitUntilInstanceRunning(&waitInput)
+				if *diOutput.Reservations[0].Instances[0].State.Name != "stopped" {
+					awschan <- "running"
+					svc.WaitUntilInstanceStopped(&waitInput)
+				}
+			}
+			if obj.State == "terminated" {
+				svc.WaitUntilInstanceExists(&waitInput)
+				if len(diOutput.Reservations) != 0 {
+					awschan <- "exists"
+					svc.WaitUntilInstanceTerminated(&waitInput)
+				}
+			}
+
+		}
+	}()
+	for {
+		select {
+		case event := <-obj.Events():
+			if exit, send = obj.ReadEvent(event); exit != nil {
+				return *exit
+			}
+		case state := <-awschan:
+			log.Printf(state)
+			obj.StateOK(false)
+			send = true
+			if send {
+				send = false
+				obj.Event()
+			}
+		}
+	}
+}
+
+// CheckApply method for Aws resource.
+func (obj *AwsRes) CheckApply(apply bool) (checkOK bool, err error) {
+	log.Printf("%s: CheckApply(%t)", obj, apply)
+
+	sess := session.Must(session.NewSession(&aws.Config{
+		Region: aws.String(obj.Region),
+	}))
+	svc := ec2.New(sess)
+
+	diInput := ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("tag:Name"),
+				Values: []*string{aws.String(obj.GetName())},
+			},
+			{
+				Name:   aws.String("instance-state-name"),
+				Values: []*string{aws.String("running"), aws.String("pending"), aws.String("stopped"), aws.String("stopping")},
+			},
+		},
+	}
+	diOutput, err := svc.DescribeInstances(&diInput)
```

Comment:
do we ever need to take about the `NextToken` ? https://godoc.org/github.com/aws/aws-sdk-go/service/ec2#DescribeInstancesOutput

Reply:
No, we are only interested in diOutput.Reservations[0].Instances[0]. If either array has more than one item, CheckApply will error. NextToken is for handling many instances at once as far as I can tell.